CDAP-2105: Queue introspection

jwang47 opened the pull request:
Same as https://github.com/caskdata/cdap/pull/2290 (merged).

Continued from https://github.com/caskdata/cdap/pull/2078

https://issues.cask.co/browse/CDAP-2105
http://builds.cask.co/browse/CDAP-RBT219

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueConsumer.java:
```diff
@@ -306,6 +306,11 @@ private boolean fetchFromCache(SortedMap<byte[], SimpleQueueEntry> entries, int
     return true;
   }
 
+  public static long getWritePointer(byte[] rowKey, int queueRowPrefixLenght) {
```
Don't add random static method like this to non-util type class. Move this to `QueueEntryRow` as

`getEnqueueWritePointer(byte[] rowKey, byte[] queueRowPrefix)`.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java:
```diff
@@ -316,6 +328,15 @@ public void configureGroups(Iterable<? extends ConsumerGroupConfig> groupConfigs
+   */
+  Multimap<Long, QueueBarrier> getAllBarriers() {
+    return scanBarriers(new AllPairCollector<Long, QueueBarrier>()).finishMultimap(
+      HashMultimap.<Long, QueueBarrier>create());
```
Is `HashMutlimap` guarantee sorting order? Shouldn't you use `LinkedHashMultimap`?

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java:
```diff
@@ -447,6 +487,20 @@ private int getInstanceIdFromColumn(byte[] column) {
+    Map<byte[], byte[]> columns = row.getColumns();
+    for (byte[] columnKey : columns.keySet()) {
+      long groupId = Bytes.toLong(columnKey);
+      QueueBarrier barrier = decodeBarrierInfo(row, groupId);
```
That's a bit wasteful to encode groupId again and use it to get from row again. You can add a new method:

```java
QueueBarrier decodeBarrierInfo(byte[] rowKey, byte[] columnValue);
```

and have both existing `decodeBarrierInfo` call that.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java:
```diff
@@ -0,0 +1,309 @@
+    this.queueAdmin = queueAdmin;
+    this.queueClientFactory = queueClientFactory;
+    this.txClient = txClient;
+    this.zkClientService = zkClientService;
```
I don't see the zkClient is used anywhere. What is it for?

jwang47 replied:
I remember something needed zkClientService to be started.. I can check what it is.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java:
```diff
@@ -0,0 +1,309 @@
+    int currentSection = 1;
+    for (Long groupId : barriers.keySet()) {
+      LOG.info("Scanning barriers for group {}", groupId);
+      Collection<QueueBarrier> groupBarriers = barriers.get(groupId);
```
You can loop over `barries.entries()`.

jwang47 replied:
Not sure how I do that and still maintain previous and current `QueueBarrier` cleanly. Seems I'd have to check every time if the `groupId` changed and do some logic to reset previous and current `QueueBarrier`.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java:
```diff
@@ -0,0 +1,309 @@
+
+      QueueBarrier previous;
+      QueueBarrier current = null;
+      while (barrierIterator.hasNext()) {
```
You should be able to use a for-each loop instead of the iterator.

jwang47 replied:
For-each loop is possible, but I didn't like it b/c I'm handling the last barrier outside the for-loop, like so:

```

      LOG.info("Scanning section {}/{}...", currentSection, barriers.size());
      stats = QueueStatistics.of(stats, scanQueue(tx, queueName, lastBarrier, null));
      LOG.info("Scanning complete", currentSection, barriers.size());
```

jwang47 replied:
Though maybe `Multimap`/`LinkedMultimap` is the wrong class to use here, since I need the entries to be ordered per key. Maybe `Map<Long, List>` would be better?

chtyim replied:
You don't need order by key (group ID), you just need to maintain the ordering of the barriers with the same order of insertion (scanning from HBase gives you from earliest to latest barrier).

jwang47 commented:
Going to address comments then reopen against hotfix/temp-fix-flow-change-instances branch.

jwang47 commented:
Reopened against hotfix/temp-fix-flow-change-instances branch, squashed commits: https://github.com/caskdata/cdap/pull/2290 (you can still review how I addressed comments in this PR)

jwang47 commented:
Merging after build passes.
